src/openfermion/utils/_bch_expansion.py

```diff
@@ -0,0 +1,157 @@
+#   limitations under the License.
+
+"""Module to efficiently compute the Baker-Campbell-Hausdorff formula."""
+
+import itertools
+import numpy as np
+from scipy.misc import comb, factorial
+
+from openfermion.utils import commutator
+
+def bch_expand(x, y, order):
+    """Compute log[e^x e^y] using the Baker-Campbell-Hausdorff formula
+    Args:
+        x: An operator for which multiplication and addition are supported.
+            For instance, a QubitOperator, FermionOperator or numpy array.
+        y: The same type as x.
+        order(int): The order to truncate the BCH expansions.
+
+    Returns:
+        z: The truncated BCH operator.
+
+    Raises:
+        ValueError: operator x is not same type as operator y.
+        ValueError: invalid order parameter.
+        ValueError: order exceeds maximum order supported.
+    """
+
+    max_order = 11
+    if (not isinstance(order, int)) or order < 0:
+        raise ValueError('Invalid order parameter.')
+    if order > max_order:
+        raise ValueError('Order exceeds maximum order supported.')
+    if type(x) != type(y):
+        raise ValueError('Operator x is not same type as operator y.')
+
+    z = None
+    terms, coeff = generate_nested_commutator(order)
+    for bin_str, c in zip(terms, coeff):
+        t = bin_str_to_commutator(bin_str, x, y)
+        if z is None:
+            z = t * c
+        else:
+            z += t * c
+
+    # Return.
+    return z
+
+def bin_str_to_commutator(bin_str, x, y):
+    char_to_xy = lambda char: x if char == '0' else y
+    op = char_to_xy(bin_str[0])
+
+    if len(bin_str) == 1:
+        return op
+    else:
+        return commutator(op, bin_str_to_commutator(bin_str[1:], x, y))
+
+def generate_nested_commutator(order):
+    """
+    using bin strings to encode nested commutators like [X,[Y,[X, ...]]] as '010...'
+    """
+    terms = []
+    coeff = []
+
+    for i in range(1, order + 1):
+        t = [list(x) for x in itertools.product(['0', '1'], repeat=i)]
+        if i > 1:
+            t = filter(lambda x: x[-1] != x[-2], t)
+        t = ["".join(x) for x in t]
+        terms += t
+
+    for t in terms:
+        split_bin_str = split_by_descending_edge(t)
+        coeff.append(compute_coeff(split_bin_str))
+    return terms, coeff
+
+def split_by_descending_edge(bin_str):
+    prev = '0'
+    split_idx = [0]
+
+    for idx, i in enumerate(bin_str):
+        if prev == '1' and i == '0':
+            split_idx.append(idx)
+        prev = i
+
+    if len(split_idx) == 1:
+        return [bin_str]
+    else:
+        return [bin_str[i:j] for i, j in zip(split_idx, split_idx[1:]+[None])]
+
+def compute_coeff(split_bin_str):
+    N = len(''.join(split_bin_str))
+    l = len(split_bin_str) - 1
+    cn = lambda n: dfs_root(split_bin_str, n, len(split_bin_str))
+    c = sum([(-1)**(n+1) / n * cn(n) for n in range(l+1, N+1)])
+    return c/N
+
+def dfs_root(split_bin_str, n, l):
```

Comment:
A better name for this function would be good. Not sure what "dfs" is. Depth first search?